# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/MediaConduitInterface.h
# Commit: 49323e9c1054
# Full Hash: 49323e9c105493ba926196802015767390304286
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 2
# Description:
#   Bug 1654112 - Use a ShutdownBlockingTicket in WebRtcCallWrapper. r=jib,bwc
#   
#   Peer connections are cleaned up on shutdown through an xpcom-shutdown observer.
#   It doesn't block shutdown throughout async operations.
#   
# ==============================================================================

diff -r 31963b9b2342 -r 49323e9c1054 dom/media/webrtc/libwebrtcglue/MediaConduitInterface.h
--- a/dom/media/webrtc/libwebrtcglue/MediaConduitInterface.h	Tue Feb 16 17:44:00 2021 +0100
+++ b/dom/media/webrtc/libwebrtcglue/MediaConduitInterface.h	Tue Feb 16 18:03:06 2021 +0100
@@ -13,6 +13,7 @@
 #include "jsapi/PeerConnectionCtx.h"
 #include "jsapi/RTCStatsReport.h"
 #include "MediaConduitErrors.h"
+#include "mozilla/media/MediaUtils.h"
 #include "mozilla/RefCounted.h"
 #include "TaskQueueWrapper.h"
 #include "VideoTypes.h"
@@ -246,15 +247,18 @@
 
   static RefPtr<WebRtcCallWrapper> Create(
       const dom::RTCStatsTimestampMaker& aTimestampMaker,
+      UniquePtr<media::ShutdownBlockingTicket> aShutdownTicket,
       SharedWebrtcState* aSharedState, webrtc::WebRtcKeyValueConfig* aTrials) {
     auto current = TaskQueueWrapper::MainAsCurrent();
-    return Create(aTimestampMaker, aSharedState->GetModuleThread(),
+    return Create(aTimestampMaker, std::move(aShutdownTicket),
+                  aSharedState->GetModuleThread(),
                   aSharedState->mAudioStateConfig,
                   aSharedState->mAudioDecoderFactory, aTrials);
   }
 
   static RefPtr<WebRtcCallWrapper> Create(
       const dom::RTCStatsTimestampMaker& aTimestampMaker,
+      UniquePtr<media::ShutdownBlockingTicket> aShutdownTicket,
       webrtc::SharedModuleThread* aModuleThread,
       const webrtc::AudioState::Config& aAudioStateConfig,
       webrtc::AudioDecoderFactory* aAudioDecoderFactory,
@@ -272,7 +276,8 @@
     return new WebRtcCallWrapper(
         aAudioDecoderFactory, std::move(videoBitrateAllocatorFactory),
         WrapUnique(webrtc::Call::Create(config, aModuleThread)),
-        std::move(eventLog), std::move(taskQueueFactory), aTimestampMaker);
+        std::move(eventLog), std::move(taskQueueFactory), aTimestampMaker,
+        std::move(aShutdownTicket));
   }
 
   static RefPtr<WebRtcCallWrapper> Create(UniquePtr<webrtc::Call> aCall) {
@@ -317,11 +322,13 @@
 
   // Allow destroying the Call instance on the Call worker thread.
   //
+  // Note that shutdown is blocked until the Call instance is destroyed.
+  //
   // This CallWrapper is designed to be sharable, and is held by several objects
   // that are cycle-collectable. TaskQueueWrapper that the Call instances use
   // for worker threads are based off SharedThreadPools, and will block
   // xpcom-shutdown-threads until destroyed. The Call instance however will hold
-  // on to its worker threads until destroyed itself.
+  // on to its worker threads until destruction.
   //
   // If the last ref to this CallWrapper is held to cycle collector shutdown we
   // end up in a deadlock where cycle collector shutdown is required to destroy
@@ -334,6 +341,7 @@
     MOZ_ASSERT(NS_IsMainThread());
     auto current = TaskQueueWrapper::MainAsCurrent();
     mCall = nullptr;
+    mShutdownTicket = nullptr;
   }
 
   const dom::RTCStatsTimestampMaker& GetTimestampMaker() const {
@@ -350,8 +358,10 @@
                     UniquePtr<webrtc::Call> aCall,
                     UniquePtr<webrtc::RtcEventLog> aEventLog,
                     UniquePtr<webrtc::TaskQueueFactory> aTaskQueueFactory,
-                    const dom::RTCStatsTimestampMaker& aTimestampMaker)
+                    const dom::RTCStatsTimestampMaker& aTimestampMaker,
+                    UniquePtr<media::ShutdownBlockingTicket> aShutdownTicket)
       : mTimestampMaker(aTimestampMaker),
+        mShutdownTicket(std::move(aShutdownTicket)),
         mAudioDecoderFactory(std::move(aAudioDecoderFactory)),
         mVideoBitrateAllocatorFactory(std::move(aVideoBitrateAllocatorFactory)),
         mEventLog(std::move(aEventLog)),
@@ -367,6 +377,7 @@
   // collisions.
   std::set<MediaSessionConduit*> mConduits;
   dom::RTCStatsTimestampMaker mTimestampMaker;
+  UniquePtr<media::ShutdownBlockingTicket> mShutdownTicket;
 
  public:
   const RefPtr<webrtc::AudioDecoderFactory> mAudioDecoderFactory;